src/primitives/poseidon.rs

```diff
@@ -0,0 +1,361 @@
+
+    /// The type used to hold duplex sponge state.
+    ///
+    /// This must be an array of length equal to the rate of the duplex sponge (allowing
+    /// for a capacity consistent with this specification's security level), that defaults
+    /// to `[None; RATE]`.
+    type Rate: Default + AsRef<[Option<F>]> + AsMut<[Option<F>]>;
+
+    /// The arity of this specification.
+    fn arity() -> usize;
+
+    /// The number of full rounds for this specification.
+    fn full_rounds() -> usize;
+
+    /// The number of partial rounds for this specification.
+    fn partial_rounds() -> usize;
+
+    /// The S-box for this specification.
+    fn sbox(val: F) -> F;
+
+    /// Side-loaded index of the first correct and secure MDS that will be generated by
+    /// the reference implementation.
+    ///
+    /// This is used by the default implementation of [`Spec::constants`]. If you are
+    /// hard-coding the constants, you may leave this unimplemented.
+    fn secure_mds(&self) -> usize;
+
+    /// Generates `(round_constants, mds, mds^-1)` corresponding to this specification.
+    fn constants(&self) -> (Vec<Self::State>, Vec<Self::State>, Vec<Self::State>) {
+        let t = Self::arity();
+        let r_f = Self::full_rounds();
+        let r_p = Self::partial_rounds();
+
+        let mut grain = grain::Grain::new(SboxType::Pow, t as u16, r_f as u16, r_p as u16);
+
+        let round_constants = (0..(r_f + r_p))
+            .map(|_| {
+                let mut rc_row = Self::State::default();
+                for (rc, value) in rc_row
+                    .as_mut()
+                    .iter_mut()
+                    .zip((0..t).map(|_| grain.next_field_element()))
+                {
+                    *rc = value;
+                }
+                rc_row
+            })
+            .collect();
+
+        let (mds, mds_inv) = mds::generate_mds(&mut grain, t, self.secure_mds());
+
+        (
+            round_constants,
+            mds.into_iter()
+                .map(|row| {
+                    let mut mds_row = Self::State::default();
+                    for (entry, value) in mds_row.as_mut().iter_mut().zip(row.into_iter()) {
+                        *entry = value;
+                    }
+                    mds_row
+                })
+                .collect(),
+            mds_inv
+                .into_iter()
+                .map(|row| {
+                    let mut mds_row = Self::State::default();
+                    for (entry, value) in mds_row.as_mut().iter_mut().zip(row.into_iter()) {
+                        *entry = value;
+                    }
+                    mds_row
+                })
+                .collect(),
+        )
+    }
+}
+
+/// Runs the Poseidon permutation on the given state.
+fn permute<F: FieldExt, S: Spec<F>>(
+    state: &mut S::State,
+    mds: &[S::State],
+    round_constants: &[S::State],
+) {
+    // TODO: Check what should happen for odd number of full rounds.
```

Comment:
The answer is that R<sub>F</sub> must always be even by construction; in the paper it's defined as R<sub>F</sub> = 2R<sub>f</sub>.